README.md
[![CRAN version](http://www.r-pkg.org/badges/version/matrixTests)](https://cran.r-project.org/package=matrixTests)
[![Build Status](https://travis-ci.com/karoliskoncevicius/matrixTests.svg?branch=master)](https://travis-ci.com/karoliskoncevicius/matrixTests)
[![codecov](https://codecov.io/gh/karoliskoncevicius/matrixTests/branch/master/graph/badge.svg)](https://codecov.io/gh/karoliskoncevicius/matrixTests)
[![dependencies](https://tinyverse.netlify.com/badge/matrixTests)](https://CRAN.R-project.org/package=matrixTests)
[![Monthly Downloads](https://cranlogs.r-pkg.org/badges/matrixTests)](https://cranlogs.r-pkg.org/badges/matrixTests)

# Matrix Tests #

A package dedicated to running multiple statistical hypothesis tests on rows and columns of matrices.

![illustration](http://karolis.koncevicius.lt/data/matrixtests/illustration.png)

## Goals ##

1. Fast execution via vectorization.
2. Convenient and detailed output format.
3. Compatibility with tests implemented in base R.
4. Careful handling of missing values and edge cases.

## Examples ##

#### 1. Bartlett's test on columns ####

Bartlett's test on every column of iris dataset using Species as groups:

```r
col_bartlett(iris[,-5], iris$Species)
```
```
             obs.tot obs.groups var.pooled df statistic                pvalue
Sepal.Length     150          3 0.26500816  2 16.005702 0.0003345076070163084
Sepal.Width      150          3 0.11538776  2  2.091075 0.3515028004158132768
Petal.Length     150          3 0.18518776  2 55.422503 0.0000000000009229038
Petal.Width      150          3 0.04188163  2 39.213114 0.0000000030547839322
```

#### 2. Welch t-test on rows ####

Welch t-test performed on each row of 2 large (million row) matrices:

```r
X <- matrix(rnorm(10000000), ncol = 10)
Y <- matrix(rnorm(10000000), ncol = 10)

row_t_welch(X, Y)  # running time: 2.4 seconds
```

Confidence interval computations can be turned-off for further increase in speed:

```r
row_t_welch(X, Y, conf.level = NA)  # running time: 1 second
```

## Available Tests ##

|           Variant                |           Name                        |           Function                 |
|----------------------------------|---------------------------------------|------------------------------------|
| **Location tests (1 group)**     | Single sample Student's t.test        | `row_t_onesample`                  |
|                                  | Single sample Wilcoxon's test         | `row_wilcoxon_onesample`           |
| **Location tests (2 groups)**    | Equal variance Student's t.test       | `row_t_equalvar`                   |
|                                  | Welch adjusted Student's t.test       | `row_t_welch`                      |
|                                  | Two sample Wilcoxon's test            | `row_wilcoxon_twosample`           |
| **Location tests (paired)**      | Paired Student's t.test               | `row_t_paired`                     |
|                                  | Paired Wilcoxon's test                | `row_wilcoxon_paired`              |
| **Location tests (2+ groups)**   | Equal variance oneway anova           | `row_oneway_equalvar`              |
|                                  | Welch's oneway anova                  | `row_oneway_welch`                 |
|                                  | Kruskal-Wallis test                   | `row_kruskalwallis`                |
|                                  | van der Waerden's test                | `row_waerden`                      |
| **Scale tests (2 groups)**       | F variance test                       | `row_f_var`                        |
| **Scale tests (2+ groups)**      | Bartlett's test                       | `row_bartlett`                     |
|                                  | Fligner-Killeen test                  | `row_flignerkilleen`               |
|                                  | Levene's test                         | `row_levene`                       |
|                                  | Brown-Forsythe test                   | `row_brownforsythe`                |
| **Association tests**            | Pearson's correlation test            | `row_cor_pearson`                  |
| **Periodicity tests**            | Cosinor                               | `row_cosinor`                      |
| **Distribution tests**           | Kolmogorov-Smirnov test               | `row_kolmogorovsmirnov_twosample`  |
| **Normality tests**              | Jarque-Bera test                      | `row_jarquebera`                   |
|                                  | Anderson-Darling test                 | `row_andersondarling`              |


## Further Information ##

For more information please refer to the [Wiki](https://github.com/karoliskoncevicius/matrixTests/wiki) page:

1. [Installation Instructions](https://github.com/karoliskoncevicius/matrixTests/wiki/Installation)
2. [Design Decisions](https://github.com/karoliskoncevicius/matrixTests/wiki/Design-Decisions)
3. [Speed Benchmarks](https://github.com/karoliskoncevicius/matrixTests/wiki/Benchmarks)
4. [Bug Fixes and Improvements to Base R](https://github.com/karoliskoncevicius/matrixTests/wiki/Bug-Fixes-and-Improvements-to-Base-R)


## See Also ##

### Literature ###

**Computing thousands of test statistics simultaneously in R**, *Holger Schwender, Tina Müller*.\
Statistical Computing & Graphics. Volume 18, No 1, June 2007.

### Packages ###

CRAN:

1. `ttests()` in the [**Rfast**](https://CRAN.R-project.org/package=Rfast) package.
2. `row.ttest.stat()` in the [**metaMA**](https://CRAN.R-project.org/package=metaMA) package.
3. `MultiTtest()` in the [**ClassComparison**](https://CRAN.R-project.org/package=ClassComparison) package.
4. `bartlettTests()` in the [**heplots**](https://CRAN.R-project.org/package=heplots) package.
5. `harmonic.regression()` in the [**HarmonicRegression**](https://CRAN.R-project.org/package=HarmonicRegression) package.

BioConductor:

1. `lmFit()` in the [**limma**](https://bioconductor.org/packages/release/bioc/html/limma.html) package.
2. `rowttests()` in the [**genefilter**](https://bioconductor.org/packages/release/bioc/html/genefilter.html) package.
3. `mt.teststat()` in the [**multtest**](https://www.bioconductor.org/packages/release/bioc/html/multtest.html) package.
4. `row.T.test()` in the [**HybridMTest**](https://www.bioconductor.org/packages/release/bioc/html/HybridMTest.html) package.
5. `rowTtest()` in the [**viper**](https://bioconductor.org/packages/release/bioc/html/viper.html) package.
6. `lmPerGene()` in the [**GSEAlm**](https://www.bioconductor.org/packages/release/bioc/html/GSEAlm.html) package.

GitHub:

1. `rowWilcoxonTests()` in the [**sanssouci**](https://github.com/pneuvial/sanssouci) package.
2. `matrix.t.test()` in the [**pi0**](https://github.com/gitlongor/pi0) package.
3. `wilcoxauc()` in the [**presto**](https://github.com/immunogenomics/presto) package.


NEWS
Package: matrixTests
====================

Version 0.2.3 [2023 Oct 05]

INTERNAL:

 * Kolmogorov-Smirnov test now works on R-devel: fix requested by CRAN after a change of argument names in `psmirnov`.


Version 0.2.2 [2023 May 1]

NEW FEATURES:

 * Added two-sample Kolmogorov-Smirnov test: `row_kolmogorovsmirnov_twosample` and `col_kolmogorovsmirnov_twosample`.


Version 0.2.1 [2023 Apr 20]

INTERNAL:

 * Slight speed improvements of various tests in cases with large sample sizes.


Version 0.2.0 [2023 Apr 07]

NEW FEATURES:

 * Added Anderson-Darling test: `row_andersondarling` and `col_andersondarling`.
 * Setting conf.level to NA now turns off confidence intervals and improves computation speed.

BUG FIXES:

 * Wilcoxon two-sample test no longer causes integer overflow on large sample sizes.
 * Waerden and Flinger-Killeen tests no longer produce NA results when a whole group has only NA values in x.
 * Wilcoxon tests no longer produce double warning messages when some rows have no remaining observations.
 * Paired t-test no longer returns var.diff = 0 when all paired observations are missing.
 * Warnings are now more robust in correctly deciding between row and column messages.

CHANGES:

 * Null-mean and null-ratio arguments are now more consistent across functions: previously named `mu` and `ratio`, now consistently named `null`.
 * The position of `null` argument is now more consistent across functions: `null` always comes before the `alternative` argument.
 * `row_ievora` and `col_ievora` functions are removed. ievora method is a simple application of a Bartlett's test followed by a Student's t-test and can be performed manually.
 * `Inf` values in `wilcox_*` tests are no longer removed but treated as highest ranks in order to be consistent with base R 4.0.0. implementation.
 * Cosinor gets a separate warning for the case where all observations come from the same time-point within the specified period.
 * Missing argument errors no longer reference `force()` in their error messages.
 * Warning message about constant values in rows are now consistent between tests.
 * Warning message about too few remaining observations are now more consistent between wilcoxon and t tests.
 * When a test cannot be executed all test-specific outputs (test statistic, standard error, and degrees of freedom, not just p-values and confidence intervals as before) are now turned to NA for that row/column.

INTERNAL:

 * Slight speed improvements of all tests in cases with sample sizes bigger than 100.
 * Improved the speed of Wilcox tests, specially in cases with ties.
 * van der Waerden test is now aligned and tested against `PMCMRplus::vanWaerdenTest()` (previously `PMCMR::vanWaerden.test()`).
 * matrixTests no longer depends on `testthat` for unit tests.
 * All unit tests are rewritten.


Version 0.1.9 [2020 Apr 30]

NEW FEATURES:

 * Added van der Waerden test: `row_waerden` and `col_waerden`.

DOCUMENTATION:

 * Changed NEWS format to markdown.


Version 0.1.8 [2020 Jan 08]

INTERNAL:

 * Adjusted unit tests to conform with changes in R-devel 4.0.0 for CRAN.


Version 0.1.7 [2019 Nov 05]

NEW FEATURES:

 * Added Cosinor test: `row_cosinor` and `col_cosinor`.


Version 0.1.6 [2019 Sep 19]

NEW FEATURES:

 * Added Levene test: `row_levene` and `col_levene`.
 * Added Brown-Forsythe test: `row_brownforsythe` and `col_brownforsythe`.


Version 0.1.5 [2019 Jul 09]

NEW FEATURES:

 * Added F-test for variance: `row_f_var` and it's column equivalent.

INTERNAL:

 * Various speed improvements across all types of tests.
 * Added missing tests for warnings about NA values in "group" argument.


Version: 0.1.4 [2019 Jun 25]

INTERNAL:

 * Improved the speed of Wilcoxon and Kruskal Wallis tests.
 * Improved the speed of Oneway and Bartlett tests.
 * Added run time benchmarks.
 * Expanded test coverage to include all lines within code at least once.
 * Added more tests for ties in `row_kruskalwallis`.


Version: 0.1.3 [2019 Jun 10]

NEW FEATURES:

 * Added Wilcoxon tests: `row_wilcox_onesample`, `row_wilcox_paired` and `row_wilcox_twosample` plus column version equivalents.

BUG FIXES:

 * Warning messages for column-wise functions no longer warn about rows and specifies columns instead.

DOCUMENTATION:

 * Small adjustments to some documentation files (spelling mistakes, ordering, etc).


Version: 0.1.2 [2019 Apr 05]

NEW FEATURES:

 * Added Jarque Bera test: `row_jarquebera` and `col_jarquebera`.
 * Added Fligner-Killeen test: `row_flignerkilleen` and `col_flignerkilleen`.

BUG FIXES:

 * The warning messages are no longer repeated 3 times in cases when the call has a namespace (i.e. `matrixTests::row_bartlett`).


Version: 0.1.1 [2019 Mar 15]

CHANGES:

 * `row_ievora` and `col_ievora` "g" (group) parameter was renamed to "b" (binary).

INTERNAL:

 * Fixed an indexing bug in `row_cor_pearson()` monte-carlo unit testing step.

DOCUMENTATION:

 * Improved help documentation.


Version: 0.1.0 [2018 Mar 5]

NEW FEATURES:

 * Created and submitted to CRAN.



R/ttest.R
#' t-test
#'
#' Performs a t-test on each row/column of the input matrix.
#'
#' Functions to perform one sample and two sample t-tests for rows/columns of matrices.
#' Main arguments and results were intentionally matched to the \code{t.test()}
#' function from default stats package. Other arguments were split into separate
#' functions:
#'
#' \code{row_t_onesample(x)} - one sample t-test on rows.
#' \code{col_t_onesample(x)} - one sample t-test on columns.
#'
#' Results should be the same as running \code{t.test(x)}
#' on every row (or column) of \code{x}.
#'
#' \code{row_t_equalvar(x, y)} - two sample equal variance t-test on rows.
#' \code{col_t_equalvar(x, y)} - two sample equal variance t-test on columns.
#'
#' Results should be the same as running \code{t.test(x, y, var.equal=TRUE)}
#' on every row (or column) of \code{x} and \code{y}.
#'
#' \code{row_t_welch(x, y)} - two sample t-test with Welch correction on rows.
#' \code{col_t_welch(x, y)} - two sample t-test with Welch correction on columns.
#'
#' Results should be the same as running \code{t.test(x, y)}
#' on every row (or column) of \code{x} and \code{y}.
#'
#' \code{row_t_paired(x, y)} - two sample paired t-test on rows.
#' \code{col_t_paired(x, y)} - two sample paired t-test on columns.
#'
#' Results should be the same as running \code{t.test(x, y, paired=TRUE)}
#' on every row (or column) of \code{x} and \code{y}.
#'
#' @param x numeric matrix.
#' @param y numeric matrix for the second group of observations.
#' @param null true values of the means for the null hypothesis.
#' A single number or numeric vector with values for each observation.
#' @param alternative alternative hypothesis to use for each row/column of x.
#' A single string or a vector with values for each observation.
#' Values must be one of "two.sided" (default), "greater" or "less".
#' @param conf.level confidence levels used for the confidence intervals.
#' A single number or a numeric vector with values for each observation.
#' All values must be in the range of [0:1] or NA.
#'
#' @return a data.frame where each row contains the results of a t.test
#' performed on the corresponding row/column of x.
#' The columns will vary depending on the type of test performed.\cr\cr
#' They will contain a subset of the following information:\cr
#' 1. obs.x - number of x observations\cr
#' 2. obs.y - number of y observations\cr
#' 3. obs.tot - total number of observations\cr
#' 4. obs.paired - number of paired observations (present in x and y)\cr
#' 5. mean.x - mean estiamte of x\cr
#' 6. mean.y - mean estiamte of y\cr
#' 7. mean.diff - mean estiamte of x-y difference\cr
#' 8. var.x - variance estiamte of x\cr
#' 9. var.y - variance estiamte of y\cr
#' 10. var.diff - variance estiamte of x-y difference\cr
#' 11. var.pooled - pooled variance estimate of x and y\cr
#' 12. stderr - standard error\cr
#' 13. df - degrees of freedom\cr
#' 14. statistic - t statistic\cr
#' 15. pvalue - p-value\cr
#' 16. conf.low - lower bound of the confidence interval\cr
#' 17. conf.high - higher bound of the confidence interval\cr
#' 18. mean.null - mean of the null hypothesis\cr
#' 19. alternative - chosen alternative hypothesis\cr
#' 20. conf.level - chosen confidence level
#'
#' @note
#' For a marked increase in computation speed turn off the calculation of
#' confidence interval by setting \code{conf.level} to NA.
#'
#' @seealso \code{t.test()}
#'
#' @examples
#' X <- iris[iris$Species=="setosa",1:4]
#' Y <- iris[iris$Species=="virginica",1:4]
#' col_t_welch(X, Y)
#'
#' # same row using different confidence levels
#' col_t_equalvar(X[,c(1,1,1)], Y[,c(1,1,1)], conf.level=c(0.9, 0.95, 0.99))
#'
#' @author Karolis Koncevičius
#' @name ttest
#' @export
row_t_equalvar <- function(x, y, null=0, alternative="two.sided", conf.level=0.95) {
  is.null(x)
  is.null(y)

  if(is.vector(x))
    x <- matrix(x, nrow=1)
  if(is.vector(y))
    y <- matrix(y, nrow=1)

  if(is.data.frame(x) && all(sapply(x, is.numeric)))
    x <- data.matrix(x)
  if(is.data.frame(y) && all(sapply(y, is.numeric)))
    y <- data.matrix(y)

  assert_numeric_mat_or_vec(x)
  assert_numeric_mat_or_vec(y)

  if(nrow(y)==1 & nrow(x)>1) {
    y <- matrix(y, nrow=nrow(x), ncol=ncol(y), byrow=TRUE)
  }

  assert_equal_nrow(x, y)

  if(length(alternative)==1)
    alternative <- rep.int(alternative, nrow(x))
  assert_character_vec_length(alternative, 1, nrow(x))

  choices <- c("two.sided", "less", "greater")
  alternative <- choices[pmatch(alternative, choices, duplicates.ok=TRUE)]
  assert_all_in_set(alternative, choices)

  if(length(null)==1)
    null <- rep.int(null, nrow(x))
  assert_numeric_vec_length(null, 1, nrow(x))
  assert_all_in_closed_interval(null, -Inf, Inf)

  if(all(is.na(conf.level)))
    conf.level[] <- NA_real_
  if(length(conf.level)==1)
    conf.level <- rep.int(conf.level, nrow(x))
  assert_numeric_vec_length(conf.level, 1, nrow(x))
  assert_all_in_closed_interval(conf.level, 0, 1, na.allow=TRUE)


  mxs  <- rowMeans(x, na.rm=TRUE)
  mys  <- rowMeans(y, na.rm=TRUE)
  mxys <- mxs - mys

  nxs  <- ncol(x) - matrixStats::rowCounts(x, value=NA)
  nys  <- ncol(y) - matrixStats::rowCounts(y, value=NA)
  nxys <- nxs + nys

  vxs <- rowVars(x, n=nxs, m=mxs, na.rm=TRUE)
  vys <- rowVars(y, n=nys, m=mys, na.rm=TRUE)

  dfs  <- nxs + nys - 2

  vs <- rep.int(0, nrow(x))
  inds <- nxs > 1
  vs[inds] <- vs[inds] + (nxs[inds]-1) * vxs[inds]
  inds <- nys > 1
  vs[inds] <- vs[inds] + (nys[inds]-1) * vys[inds]
  vs <- vs/dfs

  stders <- sqrt(vs * (1/nxs + 1/nys))

  tres <- do_ttest(mxys, null, stders, alternative, dfs, conf.level)


  w1 <- nxys < 3
  showWarning(w1, 't_equalvar', 'had less than 3 total observations')

  w2 <- !w1 & nxs < 1
  showWarning(w2, 't_equalvar', 'had zero "x" observations')

  w3 <- !w1 & nys < 1
  showWarning(w3, 't_equalvar', 'had zero "y" observations')

  w4 <- stders <= 10 * .Machine$double.eps * pmax(abs(mxs), abs(mys))
  showWarning(w4, 't_equalvar', 'had essentially constant values')


  stders[w1 | w2 | w3 | w4] <- NA
  dfs[w1 | w2 | w3 | w4]    <- NA
  tres[w1 | w2 | w3 | w4,]  <- NA

  rnames <- rownames(x)
  if(!is.null(rnames)) rnames <- make.unique(rnames)
  data.frame(obs.x=nxs, obs.y=nys, obs.tot=nxys, mean.x=mxs, mean.y=mys,
             mean.diff=mxys, var.x=vxs, var.y=vys, var.pooled=vs,
             stderr=stders, df=dfs, statistic=tres[,1], pvalue=tres[,2],
             conf.low=tres[,3], conf.high=tres[,4], mean.null=null,
             alternative=alternative, conf.level=conf.level,
             stringsAsFactors=FALSE, row.names=rnames
             )
}

#' @rdname ttest
#' @export
col_t_equalvar <- function(x, y, null=0, alternative="two.sided", conf.level=0.95) {
  row_t_equalvar(t(x), t(y), null=null, alternative=alternative, conf.level=conf.level)
}


#' @rdname ttest
#' @export
row_t_welch <- function(x, y, null=0, alternative="two.sided", conf.level=0.95) {
  is.null(x)
  is.null(y)

  if(is.vector(x))
    x <- matrix(x, nrow=1)
  if(is.vector(y))
    y <- matrix(y, nrow=1)

  if(is.data.frame(x) && all(sapply(x, is.numeric)))
    x <- data.matrix(x)
  if(is.data.frame(y) && all(sapply(y, is.numeric)))
    y <- data.matrix(y)

  assert_numeric_mat_or_vec(x)
  assert_numeric_mat_or_vec(y)

  if(nrow(y)==1 & nrow(x)>1) {
    y <- matrix(y, nrow=nrow(x), ncol=ncol(y), byrow=TRUE)
  }

  assert_equal_nrow(x, y)

  if(length(alternative)==1)
    alternative <- rep.int(alternative, nrow(x))
  assert_character_vec_length(alternative, 1, nrow(x))

  choices <- c("two.sided", "less", "greater")
  alternative <- choices[pmatch(alternative, choices, duplicates.ok=TRUE)]
  assert_all_in_set(alternative, choices)

  if(length(null)==1)
    null <- rep.int(null, nrow(x))
  assert_numeric_vec_length(null, 1, nrow(x))
  assert_all_in_closed_interval(null, -Inf, Inf)

  if(all(is.na(conf.level)))
    conf.level[] <- NA_real_
  if(length(conf.level)==1)
    conf.level <- rep.int(conf.level, nrow(x))
  assert_numeric_vec_length(conf.level, 1, nrow(x))
  assert_all_in_closed_interval(conf.level, 0, 1, na.allow=TRUE)


  mxs  <- rowMeans(x, na.rm=TRUE)
  mys  <- rowMeans(y, na.rm=TRUE)
  mxys <- mxs - mys

  nxs  <- ncol(x) - matrixStats::rowCounts(x, value=NA)
  nys  <- ncol(y) - matrixStats::rowCounts(y, value=NA)
  nxys <- nxs + nys

  vxs <- rowVars(x, n=nxs, m=mxs, na.rm=TRUE)
  vys <- rowVars(y, n=nys, m=mys, na.rm=TRUE)

  stderxs <- vxs/nxs
  stderys <- vys/nys
  stders  <- stderxs + stderys
  dfs     <- stders*stders / (stderxs*stderxs/(nxs - 1) + stderys*stderys/(nys - 1))
  stders  <- sqrt(stders)

  tres <- do_ttest(mxys, null, stders, alternative, dfs, conf.level)


  w1 <- nxs < 2
  showWarning(w1, 't_welch', 'had less than 2 "x" observations')

  w2 <- !w1 & nys < 2
  showWarning(w2, 't_welch', 'had less than 2 "y" observations')

  w3 <- stders <= 10 * .Machine$double.eps * pmax(abs(mxs), abs(mys))
  showWarning(w3, 't_welch', 'had essentially constant values')

  stders[w1 | w2 | w3] <- NA
  dfs[w1 | w2 | w3]     <- NA
  tres[w1 | w2 | w3,]   <- NA

  rnames <- rownames(x)
  if(!is.null(rnames)) rnames <- make.unique(rnames)
  data.frame(obs.x=nxs, obs.y=nys, obs.tot=nxys, mean.x=mxs, mean.y=mys,
             mean.diff=mxys, var.x=vxs, var.y=vys, stderr=stders, df=dfs,
             statistic=tres[,1], pvalue=tres[,2], conf.low=tres[,3],
             conf.high=tres[,4], mean.null=null, alternative=alternative,
             conf.level=conf.level, stringsAsFactors=FALSE, row.names=rnames
             )
}

#' @rdname ttest
#' @export
col_t_welch <- function(x, y, null=0, alternative="two.sided", conf.level=0.95) {
  row_t_welch(t(x), t(y), null=null, alternative=alternative, conf.level=conf.level)
}


#' @rdname ttest
#' @export
row_t_onesample <- function(x, null=0, alternative="two.sided", conf.level=0.95) {
  is.null(x)

  if(is.vector(x))
    x <- matrix(x, nrow=1)

  if(is.data.frame(x) && all(sapply(x, is.numeric)))
    x <- data.matrix(x)

  assert_numeric_mat_or_vec(x)


  if(length(alternative)==1)
    alternative <- rep.int(alternative, nrow(x))
  assert_character_vec_length(alternative, 1, nrow(x))

  choices <- c("two.sided", "less", "greater")
  alternative <- choices[pmatch(alternative, choices, duplicates.ok=TRUE)]
  assert_all_in_set(alternative, choices)

  if(length(null)==1)
    null <- rep.int(null, nrow(x))
  assert_numeric_vec_length(null, 1, nrow(x))
  assert_all_in_closed_interval(null, -Inf, Inf)

  if(all(is.na(conf.level)))
    conf.level[] <- NA_real_
  if(length(conf.level)==1)
    conf.level <- rep.int(conf.level, nrow(x))
  assert_numeric_vec_length(conf.level, 1, nrow(x))
  assert_all_in_closed_interval(conf.level, 0, 1, na.allow=TRUE)


  nxs <- ncol(x) - matrixStats::rowCounts(x, value=NA)
  mxs <- rowMeans(x, na.rm=TRUE)
  vxs <- rowVars(x, n=nxs, m=mxs, na.rm=TRUE)

  stders <- sqrt(vxs/nxs)
  dfs <- nxs-1

  tres <- do_ttest(mxs, null, stders, alternative, dfs, conf.level)


  w1 <- nxs < 2
  showWarning(w1, 't_onesample', 'had less than 2 "x" observations')

  w2 <- !w1 & stders <= 10 * .Machine$double.eps * abs(mxs)
  showWarning(w2, 't_onesample', 'had essentially constant values')


  stders[w1 | w2]  <- NA
  dfs[w1 | w2]     <- NA
  tres[w1 | w2,]   <- NA


  rnames <- rownames(x)
  if(!is.null(rnames)) rnames <- make.unique(rnames)
  data.frame(obs=nxs, mean=mxs, var=vxs, stderr=stders, df=dfs,
             statistic=tres[,1], pvalue=tres[,2], conf.low=tres[,3],
             conf.high=tres[,4], mean.null=null, alternative=alternative,
             conf.level=conf.level, stringsAsFactors=FALSE, row.names=rnames
             )
}

#' @rdname ttest
#' @export
col_t_onesample <- function(x, null=0, alternative="two.sided", conf.level=0.95) {
  row_t_onesample(t(x), null=null, alternative=alternative, conf.level=conf.level)
}


#' @rdname ttest
#' @export
row_t_paired <- function(x, y, null=0, alternative="two.sided", conf.level=0.95) {
  is.null(x)
  is.null(y)

  if(is.vector(x))
    x <- matrix(x, nrow=1)
  if(is.vector(y))
    y <- matrix(y, nrow=1)

  if(is.data.frame(x) && all(sapply(x, is.numeric)))
    x <- data.matrix(x)
  if(is.data.frame(y) && all(sapply(y, is.numeric)))
    y <- data.matrix(y)

  assert_numeric_mat_or_vec(x)
  assert_numeric_mat_or_vec(y)

  if(nrow(y)==1 & nrow(x)>1) {
    y <- matrix(y, nrow=nrow(x), ncol=ncol(y), byrow=TRUE)
  }

  assert_equal_nrow(x, y)
  assert_equal_ncol(x, y)

  if(length(alternative)==1)
    alternative <- rep.int(alternative, nrow(x))
  assert_character_vec_length(alternative, 1, nrow(x))

  choices <- c("two.sided", "less", "greater")
  alternative <- choices[pmatch(alternative, choices, duplicates.ok=TRUE)]
  assert_all_in_set(alternative, choices)

  if(length(null)==1)
    null <- rep.int(null, nrow(x))
  assert_numeric_vec_length(null, 1, nrow(x))
  assert_all_in_closed_interval(null, -Inf, Inf)

  if(all(is.na(conf.level)))
    conf.level[] <- NA_real_
  if(length(conf.level)==1)
    conf.level <- rep.int(conf.level, nrow(x))
  assert_numeric_vec_length(conf.level, 1, nrow(x))
  assert_all_in_closed_interval(conf.level, 0, 1, na.allow=TRUE)


  xy <- x-y

  mxs  <- rowMeans(x, na.rm=TRUE)
  mys  <- rowMeans(y, na.rm=TRUE)
  mxys <- rowMeans(xy, na.rm=TRUE)

  nxs  <- ncol(x)  - matrixStats::rowCounts(x, value=NA)
  nys  <- ncol(y)  - matrixStats::rowCounts(y, value=NA)
  nxys <- ncol(xy) - matrixStats::rowCounts(xy, value=NA)

  vxs  <- rowVars(x, n=nxs, m=mxs, na.rm=TRUE)
  vys  <- rowVars(y, n=nys, m=mys, na.rm=TRUE)
  vxys <- rowVars(xy, n=nxys, m=mxys, na.rm=TRUE)

  stders <- sqrt(vxys/nxys)
  dfs <- nxys-1

  tres <- do_ttest(mxys, null, stders, alternative, dfs, conf.level)


  w1 <- nxys < 2
  showWarning(w1, 't_paired', 'had less than 2 paired observations')

  w2 <- stders <= 10 * .Machine$double.eps * abs(mxys)
  showWarning(w2, 't_paired', 'had essentially constant values')

  stders[w1 | w2] <- NA
  dfs[w1 | w2]    <- NA
  tres[w1 | w2,]  <- NA

  rnames <- rownames(x)
  if(!is.null(rnames)) rnames <- make.unique(rnames)
  data.frame(obs.x=nxs, obs.y=nys, obs.paired=nxys, mean.x=mxs, mean.y=mys,
             mean.diff=mxys, var.x=vxs, var.y=vys, var.diff=vxys,
             stderr=stders, df=dfs, statistic=tres[,1], pvalue=tres[,2],
             conf.low=tres[,3], conf.high=tres[,4], mean.null=null,
             alternative=alternative, conf.level=conf.level,
             stringsAsFactors=FALSE, row.names=rnames
             )
}

#' @rdname ttest
#' @export
col_t_paired <- function(x, y, null=0, alternative="two.sided", conf.level=0.95) {
  row_t_paired(t(x), t(y), null=null, alternative=alternative, conf.level=conf.level)
}
